balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+}
+
+// UpdateSubConnState is unused as a StateListener is always registered when
+// creating SubConns.
+func (b *pickfirstBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
+	b.logger.Errorf("UpdateSubConnState(%v, %+v) called unexpectedly", subConn, state)
+}
+
+func (b *pickfirstBalancer) Close() {
+	b.serializer.TrySchedule(func(_ context.Context) {
+		b.closeSubConns()
+		b.state = connectivity.Shutdown
+	})
+	b.serializerCancel()
+	<-b.serializer.Done()
+}
+
+// ExitIdle moves the balancer out of idle state. It can be called concurrently
+// by the idlePicker and clientConn so access to variables should be synchronized.
+func (b *pickfirstBalancer) ExitIdle() {
+	b.serializer.TrySchedule(func(_ context.Context) {
+		if b.state == connectivity.Idle {
+			b.requestConnection()
+		}
+	})
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) closeSubConns() {
+	for _, sd := range b.subConns.Values() {
+		sd.(*scData).subConn.Shutdown()
+	}
+	b.subConns = resolver.NewAddressMap()
+}
+
+// deDupAddresses ensures that each address belongs to only one endpoint.
+func deDupAddresses(endpoints []resolver.Endpoint) []resolver.Endpoint {
+	seenAddrs := resolver.NewAddressMap()
+	newEndpoints := []resolver.Endpoint{}
+
+	for _, ep := range endpoints {
+		addrs := []resolver.Address{}
+		for _, addr := range ep.Addresses {
+			if _, ok := seenAddrs.Get(addr); ok {
+				continue
+			}
+			addrs = append(addrs, addr)
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		newEndpoints = append(newEndpoints, resolver.Endpoint{
+			Addresses:  addrs,
+			Attributes: ep.Attributes,
+		})
+	}
+	return newEndpoints
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) reconcileSubConns(newEndpoints []resolver.Endpoint) {
+	// Remove old subConns that were not in new address list.
+	oldAddrs := resolver.NewAddressMap()
+	for _, k := range b.subConns.Keys() {
+		oldAddrs.Set(k, true)
+	}
+
+	// Flatten the new endpoint addresses.
+	newAddrs := resolver.NewAddressMap()
+	for _, endpoint := range newEndpoints {
+		for _, addr := range endpoint.Addresses {
+			newAddrs.Set(addr, true)
+		}
+	}
+
+	// Shut them down and remove them.
+	for _, oldAddr := range oldAddrs.Keys() {
+		if _, ok := newAddrs.Get(oldAddr); ok {
+			continue
+		}
+		val, _ := b.subConns.Get(oldAddr)
+		val.(*scData).subConn.Shutdown()
+		b.subConns.Delete(oldAddr)
+	}
+}
+
+// shutdownRemaining shuts down remaining subConns. Called when a subConn
+// becomes ready, which means that all other subConn must be shutdown.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.subConns.Set(selected.addr, selected)
```

Comment:
We loop though and delete everything including the selected subconn, then we re-add the selected subconn?

I think this could match with `closeSubConns` then and do `b.subConns = resolver.NewAddressMap()` instead of deleting everything in a loop.